vectorstores/pgvector/pgvector.go

```diff
@@ -0,0 +1,329 @@
+func (s Store) SimilaritySearch(
+	ctx context.Context,
+	query string,
+	numDocuments int,
+	options ...vectorstores.Option,
+) ([]schema.Document, error) {
+	collectionName := s.collectionName
+	opts := s.getOptions(options...)
+	if nameSpace := s.getNameSpace(opts); nameSpace != "" {
+		collectionName = nameSpace
+	}
+	scoreThreshold, err := s.getScoreThreshold(opts)
+	if err != nil {
+		return nil, err
+	}
+	filter, err := s.getFilters(opts)
+	if err != nil {
+		return nil, err
+	}
+	embedder, err := s.embedder.EmbedQuery(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	tx, err := s.conn.Begin(ctx)
+	if err != nil {
+		return nil, err
+	}
+	whereQuerys := make([]string, 0)
+	if scoreThreshold != 0 {
+		whereQuerys = append(whereQuerys, fmt.Sprintf("data.distance < %f", 1-scoreThreshold))
+	}
+	for k, v := range filter {
+		whereQuerys = append(whereQuerys, fmt.Sprintf("(data.cmetadata ->> '%s') = '%s'", k, v))
+	}
+	whereQuery := strings.Join(whereQuerys, " AND ")
+	if len(whereQuery) == 0 {
+		whereQuery = "TRUE"
+	}
+	sql := fmt.Sprintf(`SELECT
+	data.document,
+	data.cmetadata,
+	data.distance
+FROM (
+	SELECT
+		%s.*,
+		embedding <=> $1 AS distance
+	FROM
+		%s
+		JOIN %s ON %s.collection_id=%s.uuid WHERE %s.name='%s') AS data
+WHERE %s 
+ORDER BY
+	data.distance
+LIMIT %d`, s.embeddingTableName,
+		s.embeddingTableName,
+		s.collectionTableName, s.embeddingTableName, s.collectionTableName, s.collectionTableName, collectionName,
+		whereQuery, numDocuments)
+	rows, err := tx.Query(ctx, sql, pgvector.NewVector(embedder))
+	if err != nil {
+		return nil, err
+	}
+	docs := make([]schema.Document, 0)
+	for rows.Next() {
+		doc := schema.Document{}
+		if err := rows.Scan(&doc.PageContent, &doc.Metadata, &doc.Score); err != nil {
+			return nil, err
+		}
+		docs = append(docs, doc)
+	}
+	return docs, nil
+}
+
+// Close closes the connection.
+func (s Store) Close(ctx context.Context) error {
+	return s.conn.Close(ctx)
+}
+
+func (s Store) DropTables(ctx context.Context) error {
+	_, err := s.conn.Exec(ctx, fmt.Sprintf(`DROP TABLE IF EXISTS %s`, s.collectionTableName))
+	if err != nil {
+		return err
+	}
+	_, err = s.conn.Exec(ctx, fmt.Sprintf(`DROP TABLE IF EXISTS %s`, s.embeddingTableName))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s Store) RemoveCollection(ctx context.Context) error {
+	_, err := s.conn.Exec(ctx, fmt.Sprintf(`DELETE FROM %s WHERE name = '%s'`, s.collectionTableName, s.collectionName))
+	return err
+}
+
+func (s Store) createOrGetCollection(ctx context.Context) (string, error) {
+	sql := fmt.Sprintf(`INSERT INTO %s (uuid, name, cmetadata)
+		VALUES($1, $2, $3) ON CONFLICT DO NOTHING`, s.collectionTableName)
+	_, err := s.conn.Exec(ctx, sql, uuid.New().String(), s.collectionName, s.collectionMetadata)
+	if err != nil {
+		return "", err
+	}
+	sql = fmt.Sprintf(`SELECT uuid FROM %s WHERE name = $1 ORDER BY name limit 1`, s.collectionTableName)
+	rows, err := s.conn.Query(ctx, sql, s.collectionName)
```

Comment:
Can you use `QueryRow` here (and elsewhere where you only expect to scan a single row)?

Reply:
This idea comes from langchain: `https://github.com/langchain-ai/langchain/blob/v0.0.340/libs/langchain/langchain/vectorstores/_pgvector_ data_models.py#L27`, IMO that there is only one UUID for the same name in the Collection table.

Reply:
I'm not sure I follow the connection.

This code queries to extract a single value from the table, correct? I see only one `collectionUUID` which gets overwritten in every iteration of `rows.Next`.

If you're only expecting a single row, `QueryRow` will be cleaner than `Query`

Reply:
Update code in commit b4d839d7e5805f615dabb221e1b1c45adc093d26